### DISCUSSION TITLE
babel/codemod astexplorer

### DISCUSSION THREAD

**Discussion by j4k0xb** [j4k0xb/webcrack#84] [Show and tell]
[Ability to write custom babel transformers via website (#24)](https://github.com/j4k0xb/webcrack-ui/issues/24)
@Memexurer @tecnessino

Very WIP but loading types for autocompletion and transforming already works: https://codemod-explorer.netlify.app/
I'll open-source it once it's more polished

**Comment by 0xdevalias** [comment]
> yeah I wanted to do something like this for a while, also with visualizing scope and [matchers](https://github.com/codemod-js/codemod/tree/main/packages/matchers)
> 
> _Originally posted by @j4k0xb in https://github.com/j4k0xb/webcrack-ui/issues/24#issuecomment-1752043589_

Being able to easily visualise scope/etc would be really cool.

**Reply by 0xdevalias** [reply]
Here are a few 'scope visualiser' tools I've liked ideas from in the past, might be cool to see some of those features too:

- https://mazurov.github.io/escope-demo/
- https://mazurov.github.io/eslevels-demo/
- https://jsbubbl.es/

**Reply by 0xdevalias** [reply]
Another tool that I always thought would have been cool to have better support built in to AST Explorer/similar (or variations of it where relevant):

- https://estools.github.io/esquery/
  - https://github.com/estools/esquery

**Comment by j4k0xb** [comment]
Now it highlights the scope range and references/constantViolations after clicking on an identifier:
![image](https://github.com/j4k0xb/webcrack/assets/55899582/b204daf1-5268-4ab1-8d95-9e574f898257)
![image](https://github.com/j4k0xb/webcrack/assets/55899582/a7d2dd3b-315c-4076-bf23-1b28ac40bb4c)

and a function is exposed to add debug highlights in the result panel:
```ts
Literal(path) {
  highlight(path.node, 'my hover message: ' + path.type);
}
```
![image](https://github.com/j4k0xb/webcrack/assets/55899582/1af09f81-34b1-4ca7-8cde-4e6b76843024)

**Reply by 0xdevalias** [reply]
Nice!

Any plans to have clicking on the code in the top left code panel, jump to/highlight it in the top right AST panel like AST Explorer does?

![image](https://github.com/j4k0xb/webcrack/assets/753891/40f51546-7be4-4ccc-8df1-64339b6ef118)

And/or similar 'reverse highlighting' when hovering over a section of the top right AST panel, highlighting the code it relates to in the top left code panel?

![image](https://github.com/j4k0xb/webcrack/assets/753891/7be6643f-29a0-4712-89c6-8630ed2f08e9)

**Reply by j4k0xb** [reply]
sure
